lib/tasks/export_users_auth0.rake

```diff
@@ -0,0 +1,164 @@
+
+  def initialize(auth0_domain, auth0_client_id, auth0_client_secret)
+    @auth0_domain = auth0_domain
+    @auth0_client_id = auth0_client_id
+    @auth0_client_secret = auth0_client_secret
+    @auth0_api_bearer_token = request_api_bearer_token
+  end
+
+  def export_all_users
+    users = User.all
+    users.map(&method(:export_user))
+  end
+
+  def export_users(emails)
+    users = User.where(email: emails)
+    users.map(&method(:export_user))
+  end
+
+  private def export_user(user)
+    email = user.email
+    legacy_user_record = format_legacy_user_record(user)
+    puts ">>> Exporting user #{email} <<<"
+    result = auth0_api_create_user(legacy_user_record)
+
+    if result['error'] && result['response'].fetch_values('statusCode', 'error') == [409, 'Conflict']
+      puts "= user already exists, updating user info"
+      result = auth0_api_update_user(legacy_user_record)
+    end
+
+    if result['error']
+      puts "!!! ERROR exporting user: #{result['response']}"
+    else
+      puts "+ OK"
+    end
+
+    # We don't want to be rate limited
+    sleep 0.5
+    puts
+
+    result
+  end
+
+  private def request_api_bearer_token
+    puts "Retrieving bearer token"
+    json_response = HTTP.post(
+      "https://#{@auth0_domain}/oauth/token",
+      json: {
+        "client_id": @auth0_client_id,
+        "client_secret": @auth0_client_secret,
+        "audience": "https://#{@auth0_domain}/api/v2/",
+        "grant_type": "client_credentials",
+      }
+    )
+    response = JSON.parse(json_response)
+    response['access_token']
+  end
+
+  private def auth0_api_create_user(legacy_user_record)
+    json_response = HTTP.auth("Bearer #{@auth0_api_bearer_token}")
+                        .post(
+                          "https://#{@auth0_domain}/api/v2/users",
+                          json: legacy_user_record
+                        )
+    response = JSON.parse(json_response)
+    {
+      'created' => response["created_at"].present?,
+      'updated' => false,
+      'error' => response["error"].present?,
+      'response' => response.to_h,
+    }
+  end
+
+  private def auth0_api_update_user(legacy_user_record)
+    user_id = "auth0|#{legacy_user_record['user_id']}"
+
+    update_record = legacy_user_record.reject { |k| ['user_id', 'email_verified', 'verify_email', 'email'].include?(k) }
+
+    json_response = HTTP.auth("Bearer #{@auth0_api_bearer_token}")
+                        .patch(
+                          "https://#{@auth0_domain}/api/v2/users/#{user_id}",
+                          json: update_record
+                        )
+    response = JSON.parse(json_response)
+    {
+      'created' => false,
+      'updated' => response["created_at"].present?,
+      'error' => response["error"].present?,
+      'response' => response.to_h,
+    }
+  end
+
+  private def format_legacy_user_record(user)
+    {
+      'connection' => AUTH0_DB_CONNECTION_NAME,
+      'email' => "legacy_idseq|#{user.email}",
```

Comment:
What is the goal of these prefixes i.e. `legacy_idseq`, `auth0`?
Can you add some comments to the PR? OR point to a doc?

Reply:
I believe I may have to create a doc explaining this migration strategy.
We are using auth0 database user store (https://auth0.com/docs/connections/database), and that's why it is prefixed with `auth0`.
The `legacy_idseq` is just a constant defined to help identifying these profiles during automatic migration (https://auth0.com/docs/users/guides/configure-automatic-migration)

Reply:
@tfrcarvalho I managed to make it work without prefixing the email.